File: apps/chandao-cli/src/actions/bug.rs
use clap::Subcommand;
use serde_json::json;
use std::cell::RefCell;
use std::rc::Rc;

use crate::utils;
use crate::auth::AuthManager;
use crate::client::{Client, AuthenticatedClient};

macro_rules! with_auth {
    ($client:expr, $auth:expr, $body:expr) => {{
        let mut ac = $client.authenticate($auth)?;
        $body(&mut ac)
    }};
}



#[derive(Subcommand)]
pub enum BugCommands {
    /// List bugs
    List {
        /// Product ID
        #[arg(short = 'p', long)]
        product: Option<i64>,
        /// Project ID
        #[arg(short = 'j', long)]
        project: Option<i64>,
        /// Execution ID (list bugs for a specific execution)
        #[arg(short = 'e', long)]
        execution: Option<i64>,
        /// Page
        #[arg(short = 'g', long, default_value = "1")]
        page: u32,
        /// Limit
        #[arg(short = 'n', long, default_value = "20")]
        limit: u32,
    },
    /// Get bug
    Get { id: i64 },
    /// Create bug
    Create {
        /// Product ID
        #[arg(short = 'p', long)]
        product: i64,
        /// Bug title
        #[arg(short, long)]
        title: String,
        /// Assigned to
        #[arg(short = 'a', long)]
        assigned: Option<String>,
        /// Priority (1-4)
        #[arg(short = 'i', long)]
        pri: Option<u8>,
        /// Severity (1-4)
        #[arg(short = 's', long)]
        severity: Option<u8>,
        /// Type (codeerror/config/install/security/performance/standard/automation/designdefect/others)
        #[arg(short = 'y', long)]
        r#type: Option<String>,
        /// Opened build
        #[arg(short = 'b', long)]
        opened_build: Option<String>,
        /// Steps to reproduce
        #[arg(short = 'd', long)]
        desc: Option<String>,
        /// Module ID
        #[arg(short = 'm', long)]
        module: Option<i64>,
        /// Execution ID
        #[arg(short = 'e', long)]
        execution: Option<i64>,
        /// Task ID
        #[arg(long)]
        task: Option<i64>,
        /// Story ID
        #[arg(long)]
        story: Option<i64>,
        /// OS
        #[arg(long)]
        os: Option<String>,
        /// Browser
        #[arg(long)]
        browser: Option<String>,
        /// Dry run
        #[arg(long)]
        dry_run: bool,
    },
    /// Update bug
    Update {
        id: i64,
        #[arg(short, long)]
        title: Option<String>,
        #[arg(short = 'a', long)]
        assigned: Option<String>,
        #[arg(short, long)]
        status: Option<String>,
        #[arg(short = 'i', long)]
        pri: Option<u8>,
        /// Severity (1-4)
        #[arg(long)]
        severity: Option<u8>,
        /// Type (codeerror/config/install/security/performance/standard/automation/designdefect/others)
        #[arg(short = 'y', long)]
        r#type: Option<String>,
        /// Opened build
        #[arg(short = 'b', long)]
        opened_build: Option<String>,
        /// Steps to reproduce
        #[arg(short = 'd', long)]
        steps: Option<String>,
        /// Project ID
        #[arg(short = 'j', long)]
        project: Option<i64>,
        /// Execution ID (关联到执行)
        #[arg(short = 'e', long)]
        execution: Option<i64>,
        /// Story ID
        #[arg(long)]
        story: Option<i64>,
        #[arg(long)]
        dry_run: bool,
    },
    /// Resolve a bug
    Resolve {
        id: i64,
        /// Resolution: fixed, postponed, bydesign, willnotfix, duplicate, external
        #[arg(short, long)]
        resolution: String,
        #[arg(short, long)]
        comment: Option<String>,
        /// Resolved date
        #[arg(long)]
        resolved_date: Option<String>,
        /// Resolved build
        #[arg(long)]
        resolved_build: Option<String>,
        /// Assigned to
        #[arg(short = 'a', long)]
        assigned_to: Option<String>,
        #[arg(long)]
        dry_run: bool,
    },
    /// Close a bug
    Close {
        id: i64,
        #[arg(short, long)]
        comment: Option<String>,
        #[arg(long)]
        dry_run: bool,
    },
    /// Activate a bug
    Activate {
        id: i64,
        /// Opened build (影响版本)
        #[arg(short = 'b', long)]
        opened_build: Option<String>,
        /// Assigned to (激活后指派给)
        #[arg(short = 'a', long)]
        assigned_to: Option<String>,
        #[arg(short, long)]
        comment: Option<String>,
        #[arg(long)]
        dry_run: bool,
    },
    /// Delete bug
    Delete {
        id: i64,
        #[arg(long)]
        yes: bool,
        #[arg(long)]
        dry_run: bool,
    },
}


pub fn handle_bug(
    client: &Client,
    auth: &Rc<RefCell<AuthManager>>,
    cmd: &BugCommands,
) -> Result<(), String> {
    match cmd {
        BugCommands::List { product, project, execution, page, limit } => with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
            let path = if let Some(e) = execution {
                format!("/executions/{}/bugs?pageID={}&recPerPage={}", e, page, limit)
            } else if let Some(j) = project {
                format!("/projects/{}/bugs?pageID={}&recPerPage={}", j, page, limit)
            } else if let Some(p) = product {
                format!("/products/{}/bugs?pageID={}&recPerPage={}", p, page, limit)
            } else {
                format!("/bugs?pageID={}&recPerPage={}", page, limit)
            };
            let data = ac.get(&path)?;
            utils::print_table(&data, &["id", "title", "status", "severity", "pri", "assignedTo"]);
            Ok(())
        }),
        BugCommands::Get { id } => with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
            let data = ac.get(&format!("/bugs/{}", id))?;
            utils::print_json(&data);
            Ok(())
        }),
        BugCommands::Create { product, title, assigned, pri, severity, r#type, opened_build, desc, module, execution, task, story, os, browser, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 创建Bug: {}", title); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({"productID": product, "title": title});
                if let Some(t) = r#type {
                    body["type"] = json!(t);
                }
                if let Some(ob) = opened_build {
                    body["openedBuild"] = json!([ob]);
                }
                if let Some(a) = assigned { body["assignedTo"] = json!(a); }
                if let Some(p) = pri { body["pri"] = json!(p); }
                if let Some(s) = severity { body["severity"] = json!(s); }
                if let Some(d) = desc { body["steps"] = json!(d); }
                if let Some(m) = module { body["module"] = json!(m); }
                if let Some(e) = execution { body["execution"] = json!(e); }
                if let Some(t) = task { body["task"] = json!(t); }
                if let Some(s) = story { body["story"] = json!(s); }
                if let Some(o) = os { body["os"] = json!(o); }
                if let Some(b) = browser { body["browser"] = json!(b); }
                let result = ac.post("/bugs", &body)?;
                println!("✅ Bug创建成功");
                utils::print_json(&result);
                Ok(())
            })
        }
        BugCommands::Update {
            id,
            title,
            assigned,
            status,
            pri,
            severity,
            r#type,
            opened_build,
            steps,
            project,
            execution,
            story,
            dry_run,
        } => {
            if *dry_run {
                println!("🔍 [DRY-RUN] 更新Bug #{}", id);
                return Ok(());
            }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({});
                if let Some(t) = title {
                    body["title"] = json!(t);
                }
                if let Some(a) = assigned {
                    body["assignedTo"] = json!(a);
                }
                if let Some(s) = status {
                    body["status"] = json!(s);
                }
                if let Some(p) = pri {
                    body["pri"] = json!(p);
                }
                if let Some(s) = severity {
                    body["severity"] = json!(s);
                }
                if let Some(ty) = r#type {
                    body["type"] = json!(ty);
                }
                if let Some(ob) = opened_build {
                    body["openedBuild"] = json!([ob]);
                }
                if let Some(st) = steps {
                    body["steps"] = json!(st);
                }
                if let Some(j) = project {
                    body["project"] = json!(j);
                }
                if let Some(e) = execution {
                    body["execution"] = json!(e);
                }
                if let Some(s) = story {
                    body["story"] = json!(s);
                }
                let result = ac.put(&format!("/bugs/{}", id), &body)?;
                println!("✅ Bug #{} 更新成功", id);
                utils::print_json(&result);
                Ok(())
            })
        }
        BugCommands::Resolve { id, resolution, comment, resolved_date, resolved_build, assigned_to, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 解决Bug #{}: {}", id, resolution); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({"resolution": resolution});
                if let Some(c) = comment { body["comment"] = json!(c); }
                if let Some(d) = resolved_date { body["resolvedDate"] = json!(d); }
                if let Some(b) = resolved_build { body["resolvedBuild"] = json!(b); }
                if let Some(a) = assigned_to { body["assignedTo"] = json!(a); }
                ac.put(&format!("/bugs/{}/resolve", id), &body)?;
                println!("✅ Bug #{} 已解决: {}", id, resolution);
                Ok(())
            })
        }
        BugCommands::Close { id, comment, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 关闭Bug #{}", id); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({});
                if let Some(c) = comment { body["comment"] = json!(c); }
                ac.put(&format!("/bugs/{}/close", id), &body)?;
                println!("✅ Bug #{} 已关闭", id);
                Ok(())
            })
        }
        BugCommands::Activate { id, opened_build, assigned_to, comment, dry_run } => {
            if *dry_run { println!("🔍 [DRY-RUN] 激活Bug #{}", id); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                let mut body = json!({});
                if let Some(ob) = opened_build { body["openedBuild"] = json!([ob]); }
                if let Some(a) = assigned_to { body["assignedTo"] = json!(a); }
                if let Some(c) = comment { body["comment"] = json!(c); }
                ac.put(&format!("/bugs/{}/activate", id), &body)?;
                println!("✅ Bug #{} 已激活", id);
                Ok(())
            })
        }
        BugCommands::Delete { id, yes, dry_run } => {
            if !yes && !dry_run { eprintln!("⚠️  确认删除Bug #{}？使用 --yes", id); return Ok(()); }
            if *dry_run { println!("🔍 [DRY-RUN] 删除Bug #{}", id); return Ok(()); }
            with_auth!(client, auth, |ac: &mut AuthenticatedClient| {
                ac.delete(&format!("/bugs/{}", id))?;
                println!("✅ Bug #{} 已删除", id);
                Ok(())
            })
        }
    }
}

File: skills/chandao/scripts/bug.js
#!/usr/bin/env node
/**
 * chandao bug module
 *
 * Usage:
 *   node bug.js --action list [--product N] [--project N] [--execution N]
 *   node bug.js --action get --id <id>
 *   node bug.js --action create --product <id> --title <title>
 *   node bug.js --action update --id <id> --title <title>
 *   node bug.js --action resolve --id <id> --resolution <fixed|bydesign|willnotfix|external|duplicate|delay|postponed|notrepro>
 *   node bug.js --action close --id <id>
 *   node bug.js --action activate --id <id>
 *   node bug.js --action delete --id <id>
 *   node bug.js --action <write-action> --dry-run
 *   node bug.js --action delete --yes
 */

import { api, paginate } from "./auth.js";

const WRITE_ACTIONS = new Set([
  "create",
  "update",
  "resolve",
  "close",
  "activate",
  "delete",
]);

function parseArgs() {
  const args = process.argv.slice(2);
  const params = {};
  for (let i = 0; i < args.length; i++) {
    if (args[i].startsWith("--")) {
      const key = args[i].slice(2);
      params[key] = args[i + 1] ?? true;
      if (params[key] !== true) i++;
    } else if (!params.action) {
      params.action = args[i];
    }
  }
  return params;
}

function checkDryRun(action) {
  if (WRITE_ACTIONS.has(action) && process.argv.includes("--dry-run")) {
    console.log("[DRY RUN] Would execute:", process.argv.slice(2).join(" "));
    process.exit(0);
  }
}

function checkYes(action) {
  if (action === "delete" && !process.argv.includes("--yes")) {
    console.log("Delete requires --yes to confirm.");
    process.exit(0);
  }
}

// ── Actions ───────────────────────────────────────────────────────
async function list(params) {
  let path = "/bugs";
  if (params.product) path = `/products/${params.product}/bugs`;
  else if (params.project) path = `/projects/${params.project}/bugs`;
  else if (params.execution) path = `/executions/${params.execution}/bugs`;

  path += paginate("", {
    limit: params.limit,
    page: params.page,
    query: { orderBy: "id_desc" },
  });

  const { data } = await api("GET", path);
  const items = data.bugs || data.data || [];
  if (!items.length) {
    console.log("No bugs found.");
    return;
  }

  const header = `| id | title | severity | pri | status |`;
  console.log(header);
  console.log("─".repeat(header.length));
  for (const b of items) {
    console.log(
      `| ${b.id} | ${(b.title || "").slice(0, 40)} | ${b.severity || "-"} | ${b.pri || "-"} | ${b.status || "-"} |`,
    );
  }
  console.log(`\nTotal: ${items.length}`);
}

async function get(params) {
  const { data } = await api("GET", `/bugs/${params.id}`);
  console.log(JSON.stringify(data, null, 2));
}

async function create(params) {
  if (!params.product || !params.title) {
    console.error("Error: create requires --product and --title");
    process.exit(1);
  }

  const body = {
    productID: Number(params.product),
    title: params.title,
    openedBuild: params.openedBuild
      ? typeof params.openedBuild === "string"
        ? [params.openedBuild]
        : params.openedBuild
      : ["trunk"],
  };
  if (params.project) body.project = Number(params.project);
  if (params.execution) body.execution = Number(params.execution);
  if (params.severity) body.severity = Number(params.severity);
  if (params.pri) body.pri = Number(params.pri);
  if (params.type) body.type = params.type;
  if (params.steps) body.steps = params.steps;
  if (params.story) body.story = Number(params.story);

  const { data } = await api("POST", "/bugs", body);
  if (data.status === "success") {
    console.log(`Bug created successfully. ID: ${data.id}`);
  } else {
    console.error("Failed to create bug:", JSON.stringify(data));
    process.exit(1);
  }
}

async function update(params) {
  if (!params.id) {
    console.error("Error: update requires --id");
    process.exit(1);
  }

  const { data: current } = await api("GET", `/bugs/${params.id}`);
  if (current.status !== "success") {
    console.error("Failed to get bug:", JSON.stringify(current));
    process.exit(1);
  }

  const body = {};
  const fields = [
    "title",
    "severity",
    "pri",
    "type",
    "steps",
    "story",
    "assignedTo",
    "status",
    "execution",
  ];
  for (const f of fields) {
    if (params[f] !== undefined) body[f] = params[f];
    else if (current[f] !== undefined) body[f] = current[f];
  }
  // Ensure execution is numeric if provided
  if (params.execution) body.execution = Number(params.execution);
  else if (current.execution) body.execution = Number(current.execution);

  const { data } = await api("PUT", `/bugs/${params.id}`, body);
  if (data.status === "success") {
    console.log(`Bug ${params.id} updated successfully.`);
  } else {
    console.error("Failed to update bug:", JSON.stringify(data));
    process.exit(1);
  }
}

async function resolve(params) {
  if (!params.id) {
    console.error("Error: resolve requires --id");
    process.exit(1);
  }
  if (!params.resolution) {
    console.error("Error: resolve requires --resolution");
    process.exit(1);
  }
  if (!params.resolvedBuild) {
    console.error("Error: resolve requires --resolvedBuild (影响版本)");
    process.exit(1);
  }

  const body = {
    resolution: params.resolution,
    resolvedBuild: params.resolvedBuild,
  };
  if (params.assignedTo) body.assignedTo = params.assignedTo;

  const { data } = await api("PUT", `/bugs/${params.id}/resolve`, body);
  if (data.status === "success") {
    console.log(
      `Bug ${params.id} resolved (resolution: ${params.resolution}).`,
    );
  } else {
    console.error("Failed to resolve bug:", JSON.stringify(data));
    process.exit(1);
  }
}

async function close(params) {
  const { data } = await api("PUT", `/bugs/${params.id}/close`);
  if (data.status === "success") {
    console.log(`Bug ${params.id} closed.`);
  } else {
    console.error("Failed to close bug:", JSON.stringify(data));
    process.exit(1);
  }
}

async function activate(params) {
  if (!params.openedBuild) {
    console.error("Error: activate requires --openedBuild (影响版本)");
    process.exit(1);
  }

  const openedBuild =
    typeof params.openedBuild === "string"
      ? [params.openedBuild]
      : params.openedBuild;
  const { data } = await api("PUT", `/bugs/${params.id}/activate`, {
    openedBuild,
  });
  if (data.status === "success") {
    console.log(`Bug ${params.id} activated.`);
  } else {
    console.error("Failed to activate bug:", JSON.stringify(data));
    process.exit(1);
  }
}

async function del(params) {
  const { data } = await api("DELETE", `/bugs/${params.id}`);
  if (data.status === "success") {
    console.log(`Bug ${params.id} deleted.`);
  } else {
    console.error("Failed to delete bug:", JSON.stringify(data));
    process.exit(1);
  }
}

// ── Dispatch ──────────────────────────────────────────────────────
const actions = {
  list,
  get,
  create,
  update,
  resolve,
  close,
  activate,
  delete: del,
};

async function main() {
  const params = parseArgs();
  if (!params.action || !actions[params.action]) {
    console.error(
      `Usage: node bug.js --action <${Object.keys(actions).join("|")}>`,
    );
    process.exit(1);
  }

  checkDryRun(params.action);
  checkYes(params.action);

  await actions[params.action](params);
}

main().catch((err) => {
  console.error("Fatal error:", err.message);
  process.exit(1);
});
